### ISSUE TITLE
Consider replacing Commons-RDF with Jena types

### ISSUE THREAD

**Issue by acoburn** [trellis-ldp/trellis#358]
Commons-RDF provides some nice abstractions, especially when working across multiple RDF implementations, but in practice, this seems to just add work for the garbage collector.

If we just use Jena types directly, the number of small objects will be reduced significantly. It would also mean that we could simplify some of the `*Service` interfaces: for example, the features of `IOService` (and a few others) can go directly into the core HTTP layer.

**Comment by ajs6f** [comment]
👍

**Comment by acoburn** [comment]
I've started looking into this, and I'd like a little feedback on some of the type changes that will be involved.

Wherever we have a CommonsRDF `Graph`, that can simply become a Jena `Model`.
Likewise, a CommonsRDF `Dataset` can become a Jena `DatasetGraph`.

In cases where we stream out CommonsRDF `Quad` or `Triple` objects, I think these should be Jena `Quad` and `Triple` objects (rather than `Statement`, since they are not connected to a particular `Model`)

Most of the remaining interface changes would involve changing CommonsRDF `IRI` to a Jena `Resource`

There are a few instances where there is a `Model`-free mapping from a BlankNode to an IRI; here, I think it might make sense to do a Jena `Node` to `Node` conversion, since these are not tied to a particular `Model`.

Otherwise, if we start using the Jena `Resource` interface, we may want to change the Trellis `Resource` to something else. How about `TrellisResource`?

**Comment by pmcb55** [comment]
Hi,

Just a quick comment on this - did you ever consider moving to RDF4J (formally known as Sesame) instead of continuing with Jena? Someone elsewhere on this repo mentioned something about the '**_legacy inertia of Jena_**' or something, and in my opinion RDF4J is a far more commercial, enterprise-grade library, e.g. it forms the basis of GraphDB from Ontotext, and what was BlazeGraph (now Amazon Neptune). It's also very actively and professionally maintained, and is laser focused on being a pure open-source Java library for RDF and SPARQL (rather than being a mismash of research-related extensions for things like OWL!).

I'm just playing devil's advocate here, so please take my tongue-in-cheek comparison of Jena vs RDF4J above with a very large grain of salt :) !

Cheers,

Pat.

**Comment by acoburn** [comment]
@pmcb55 thanks for raising this and bringing in rdf4j. I suspect that the a big part of the focus on using Jena has to do with the fact that the two main contributors on this project are Jena PMC members and we have more familiarity with Jena. That said, I am really not against using rdf4j; in fact, the API is, in a lot of ways, much nicer to work with than Jena's API. Another consideration -- perhaps you might know the answer here -- relates to JSON-LD support, particularly the new 1.1 features (framing, etc). Jena simply brings in java-jsonld for all JSON-LD processing, and so support for any of those features will depend on the underlying java-jsonld library implementing them (there has been some recent activity in java-jsonld, but I have wondered about the longterm maintenance of that codebase); do you know anything about how these features are supported (or planned to be supported) in rdf4j?

**Comment by acoburn** [comment]
While this will probably mean more work for us, I think it would make sense to make this decision based on real runtime data. We have a lot of performance data for the current CommonsRDF-based type system; what do those data look like for an RDF4J-based API and what do they look like for a Jena-based API?

I strongly suspect that RDF4J and Jena will both perform well; they are both mature libraries with plenty of time spent figuring out any performance issues. Presuming that RDF4J and Jena are both within some acceptable performance range (i.e. slightly faster, less garbage collection than right now), then I think it comes down to the ergonomics of the types. I have less familiarity with RDF4J (as compared with Jena), but I am sufficiently familiar with the API (via JavaDocs) to know that RDF4J has a much more modern feel to it.

In terms of a specific plan, I'd like to first implement a pure RDF4J I/O layer: a corollary to `trellis-io-jena`. Once that is in place, we'll need a branch with Jena types and a branch with RDF4J types. Then the testing can commence.

It might also make sense to push this off to the 0.10 release so that it doesn't block 0.9. It would be nice to work on these two branches when there aren't other changes happening to the Trellis API structure.

**Comment by pmcb55** [comment]
Hi @acoburn - firstly, I think RDF4J might also pull in java-json-ld, but I'm not sure. Their JSON-LD support has had fairly recent enhancements though (e.g. https://github.com/eclipse/rdf4j/issues/1049).

On the second issue (i.e. a performance comparison between RDF4J and Jena on basic type usage), I'd be surprised to see a noticeable difference there. But I suspect higher-order features like SPARQL performance might show big differences, as I believe a lot of work has gone into query optimization in RDF4J (possibly due to it being the backbone for commercial triplestores like GraphDB and Blazegraph).

If you don't mind I'd like to post a link to this specific issue in the RDF4J mailing list, as I strongly suspect that group would be interested in the specific questions you raise above (and I think I can guarantee a helpful and far more informed opinion from them than from me) - would you be Ok with that?

**Comment by acoburn** [comment]
Hi, yes, feel free to reference this issue in the RDF4J mailing list.

In terms of query optimization, because this is LDP, SPARQL-Update always
occurs in the context of a single LDP Resource, so unless you have a
massive resource with thousands and thousands of triples or you have a very
peculiar update query (not impossible, though it would be a bit of a
pathological example), those operations tend to be very fast, because
you've got a high level of data locality. This is very different than
optimizing for an entire triplestore to support arbitrary queries.

I would point out that the "reference implementation" for Trellis
persistence does use a Triplestore (or Jena TDB2 datastore) and there the
code does make a lot of queries across an entire Dataset, but for anyone
who wants higher performance, they should absolutely use the RDBMS- or
Cassandra-based persistence layer.

We have some baseline performance data available here:
http://drastic-testbed.umd.edu. The RDBMS version handles about 300
concurrent write operations per second (with file-based memento storage
turned on). The Cassandra version was around 500 write requests per second
in a clustered arrangement.

The other thing to mention is that, because we are currently using
CommonsRDF, we are working with an API with a very small footprint, which
in this case is very good. It also means that the code isn't doing anything
particularly complicated with those RDF types.

**Comment by namedgraph** [comment]
@pmcb55 please don't spread FUD about Jena. "mismash of research-related extensions for things like OWL" is a fluent `Resource`/`Model` API that extends all the way to ontologies, something RDF4J has no alternative for. And don't tell me people don't use ontologies.

**Comment by jeenbroekstra** [comment]
Hi all, I got alerted to this discussion by @pmcb55. I'm on the RDF4J development team. I don't think it's my place to advocate here, but if there's anything I can do to help you make an informed choice in this space (code examples, questions about features/roadmaps, etc.), do let me know.

**Comment by hmottestad** [comment]
Hi, like @jeenbroekstra I'm also a developer on the RDF4J project, though with a lot mess experience.

Unlike Jeen I'm quite happy to give my opinions from working with RDF4J and Jena over the last 5 years.

I like both, and they both have their strengths and weaknesses. Both are fairly fast, but sometimes Jena is faster and sometimes RDF4J is faster. In the previous project I was on we ended up using mostly RDF4J and some Jena. Jena was needed for getting the performance up to par for isomorphic().

We also had issues with the RDFS reasoner in RDF4J. Ended up writing a new one, since we didn't want to use the Jena one (although much faster). Writing my own reasoner and getting it integrated into the RDF4J project was much easier than I would have thought. API throughout the code is very well thought out and very consistent.

Currently I'm the developer of the SHACL support in RDF4J. It's far from full coverage of the SHACL spec, but implements the most useful features. It's also currently the only incremental SHACL engine in the world, and can efficiently scale as far as you can scale the underlying datastore with constant time validation on changes. Essentially it validates efficiently based on the changes in the transaction and does not need to validate the entire database for every change.

The open source SHACL engine for Jena that TopQuadrant implemented is geared towards validating documents, so won't be useful if you intend to add SHACL restrictions to a database. Mind you the Stardog implementation isn't incremental either, but they still advertise it as restrictions for a transactional database. My experience with their OWL based restrictions (which I assume the SHACL engine is fairly equivalent to) is that it did not scale at all. I've spent hundreds of hours carefully tweaking the restrictions to make sure that the generated SPARQL runs in close to constant time. Restrictions like "max 100" took infinite time to execute.

**Comment by ajs6f** [comment]
Just a few thoughts:

@namedgraph As you'll recall, @acoburn and I are both Jena committers, so there's not much risk of us making a quick or uninformed decision 😉 . I'm sure @pmcb55 just meant to share his own experience, and that's always useful. In fact, it's a good reminder to me that several times in the past few years we've started half-hearted conversations on the Jena lists about a new, more modern API, but we haven't pushed them forward. Perhaps it's time to move forward when committers are getting feedback on side projects that reference potential improvement for Jena's APIs, eh? 😁 

@jeenbroekstra Thank you very much for the offer of advice, and I'm sure we can make use of it.

@hmottestad Those are good data points. At this time, I don't think we have any plans that would require inference or constraint in any way; Trellis is a rather "purist" LDP implementation. But it's good to have that info, thank you!

**Comment by acoburn** [comment]
@jeenbroekstra and @hmottestad thank you both for the background and the generous offer for advice.

There is, in fact, a [ConstraintService](https://www.trellisldp.org/docs/trellis/current/apidocs/org/trellisldp/api/ConstraintService.html) in Trellis that is used to enforce some very simple structures on user-managed triples. This is primarily in place to ensure that certain LDP rules are followed (direct/indirect containers require either `ldp:hasMemberRelation` or `ldp:isMemberOfRelation` but not both, `ldp:memberRelation` must be an IRI, etc, etc). In practice, there are very few of these constraints -- only what is required for LDP. Internally, these constraint rules are loaded dynamically via the Java `ServiceLoader` so in principle, anyone who is wiring together a deployable Trellis system has some flexibility in defining what (if any) constraints apply to the RDF graphs.

Because the basic LDP constraints are so simple (and, in a sense, "static"), this is all implemented imperatively, in Java code.  On the other hand, it would be very interesting to implement this declaratively with SHACL. There may be some performance implications here, but it could make this mechanism much more flexible to have a SHACL engine involved, especially if a Trellis admin wanted to enforce certain repository-wide shapes on the user-supplied graphs.

I have also thought about writing an external service that could be used to check arbitrary constraints on arbitrary LDP resources. This would likely be similar to the [LDPath](https://github.com/trellis-ldp/camel-ldp-recipes) service that I already have.

**Comment by ajs6f** [comment]
This issue is moving in my mind from "speculative" to "technical debt". In other words, while we can debate whether Jena or RDF4J is a better choice, but I find it hard to believe that Commons RDF is really giving us the value we hoped for.

I'd like to propose that we commit to either Jena or RDF4J and close this ticket by migrating. We are getting closer every day to 1.0, and this task won't get any easier. Anyone want to speak up for Commons?

**Comment by ajs6f** [comment]
After more discussion in various fora, no one has spoken up for Commons RDF. We've established that because of our previous experience, Jena is a more natural candidate for substitution, at least for now. This is obviously a simple-but-big change, so I'm guessing it would best be done as a series of pieces in which we change one section after another of the API/SPI to use Jena directly. If that sounds good, I will open a ticket or two attempting to line out some appropriate sections.

**Comment by acoburn** [comment]
Agreed, and I would like to start working on this fairly soon. If you can identify ways to split this up into different pieces, that would be great.

**Comment by ajs6f** [comment]
I've started with one simple ticket-- let's make sure this approach makes sense before investing too far into it. Do you think a Github project would work to organize this large but simple task? We haven't done too much with them yet.

**Comment by acoburn** [comment]
We currently have https://github.com/orgs/trellis-ldp/projects/2, but feel free to add an additional project specifically for this. You should have the correct privileges to do so, but let me know if you don't.

**Comment by ajs6f** [comment]
Yeah, I'm thinking we could use something bigger than a single ticket but smaller than… everything.

https://github.com/orgs/trellis-ldp/projects/4
